File: src/core/engine.js
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { State } from './state.js';
import { Telemetry } from './telemetry.js';
import { ConnectorRegistry } from '../connectors/registry.js';
import { Correlator } from '../correlate/linker.js';
import { ReportGenerator } from '../reporting/generator.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const WORKFLOWS_PATH = join(__dirname, '..', '..', 'config', 'workflows.json');

export class Engine {
    constructor(dbPath) {
        this.state = new State(dbPath);
        this.telemetry = new Telemetry(this.state);
        this.registry = new ConnectorRegistry(this.state, this.telemetry);
        this.correlator = new Correlator(this.state);
        this.reporter = new ReportGenerator(this.state);
        this.workflows = JSON.parse(readFileSync(WORKFLOWS_PATH, 'utf-8'));
        this.running = new Map();
        this.aborted = new Set();
    }

    async investigate(name, knowns, workflow = 'person_full', config = {}) {
        const wf = this.workflows.workflows[workflow];
        if (!wf) throw new Error(`unknown workflow: ${workflow}`);

        const investigationId = this.state.createInvestigation(name, workflow, config);

        for (const known of knowns) {
            this.state.addKnown(investigationId, known.type, known.value);
        }

        this.state.updateInvestigationStatus(investigationId, 'running');
        this.running.set(investigationId, { workflow: wf, started: Date.now() });

        try {
            await this._executeWorkflow(investigationId, wf);
            this.state.updateInvestigationStatus(investigationId, 'completed');
        } catch (e) {
            this.state.updateInvestigationStatus(investigationId, 'failed');
            this.telemetry.error(investigationId, null, e.message);
            throw e;
        } finally {
            this.running.delete(investigationId);
        }

        return investigationId;
    }

    async _executeWorkflow(investigationId, workflow) {
        const completed = new Set();
        const phases = workflow.phases;

        while (completed.size < phases.length) {
            if (this.aborted.has(investigationId)) {
                throw new Error('investigation aborted');
            }

            const ready = phases.filter(p =>
                !completed.has(p.id) &&
                p.depends_on.every(dep => completed.has(dep))
            );

            if (ready.length === 0 && completed.size < phases.length) {
                throw new Error('workflow deadlock — unresolvable dependencies');
            }

            const parallel = ready.filter(p => {
                const connectors = this.registry.forPhase(p);
                return connectors.some(c => c.connector.config.parallel);
            });
            const sequential = ready.filter(p => !parallel.includes(p));

            if (parallel.length > 0) {
                await Promise.all(parallel.map(p => this._executePhase(investigationId, p)));
                for (const p of parallel) completed.add(p.id);
            }

            for (const phase of sequential) {
                await this._executePhase(investigationId, phase);
                completed.add(phase.id);
            }
        }
    }

    async _executePhase(investigationId, phase) {
        this.state.recordPhaseRun(investigationId, phase.id);
        this.telemetry.phaseStart(investigationId, phase.id);

        try {
            if (phase.internal) {
                await this._executeInternalPhase(investigationId, phase);
            } else {
                await this._executeConnectorPhase(investigationId, phase);
            }
            this.state.completePhaseRun(investigationId, phase.id, 'completed');
            this.telemetry.phaseEnd(investigationId, phase.id, 'completed');
        } catch (e) {
            this.state.completePhaseRun(investigationId, phase.id, 'failed');
            this.telemetry.phaseEnd(investigationId, phase.id, 'failed');
        }
    }

    async _executeConnectorPhase(investigationId, phase) {
        const connectors = this.registry.forPhase(phase);
        const knowns = this.state.getKnowns(investigationId);
        const tasks = [];

        if (!this._healthCache) this._healthCache = new Map();

        for (const { key, connector } of connectors) {
            if (!connector.config.native && connector.config.binary) {
                if (!this._healthCache.has(key)) {
                    const h = await connector.healthCheck();
                    this._healthCache.set(key, h.ok);
                }
                if (!this._healthCache.get(key)) continue;
            }

            const acceptedKnowns = knowns.filter(k => connector.config.accepts.includes(k.type));

            if (acceptedKnowns.length === 0 && phase.input_types) {
                const entities = this.state.getEntities(investigationId);
                const seen = new Set(acceptedKnowns.map(k => `${k.type}:${k.value}`));
                for (const entity of entities) {
                    for (const inputType of phase.input_types) {
                        const value = this._entityToInput(entity, inputType);
                        if (value && connector.config.accepts.includes(inputType)) {
                            const key2 = `${inputType}:${value}`;
                            if (!seen.has(key2)) {
                                seen.add(key2);
                                acceptedKnowns.push({ type: inputType, value });
                            }
                        }
                    }
                }
            }

            const maxInputsPerConnector = connector.config.native ? 20 : 5;
            const limited = acceptedKnowns.slice(0, maxInputsPerConnector);
            for (const known of limited) {
                tasks.push({ connector, key, inputType: known.type, inputValue: known.value });
            }
        }

        const maxParallel = 8;
        for (let i = 0; i < tasks.length; i += maxParallel) {
            const batch = tasks.slice(i, i + maxParallel);
            await Promise.allSettled(
                batch.map(t => t.connector.run(investigationId, phase.id, t.inputType, t.inputValue))
            );
        }
    }

    async _executeInternalPhase(investigationId, phase) {
        if (phase.id === 'correlation') {
            this.correlator.correlate(investigationId);
        } else if (phase.id === 'reporting') {
            await this.reporter.generate(investigationId);
        }
    }

    _entityToInput(entity, inputType) {
        const mapping = {
            email: () => entity.data.address,
            domain: () => entity.data.name,
            subdomain: () => entity.data.name,
            ip: () => entity.data.address,
            url: () => entity.data.url,
            username: () => entity.data.username,
            phone: () => entity.data.number
        };
        const fn = mapping[inputType];
        return fn ? fn() : null;
    }

    abort(investigationId) {
        this.aborted.add(investigationId);
    }

    getStatus(investigationId) {
        const investigation = this.state.getInvestigation(investigationId);
        const stats = this.state.getStats(investigationId);
        const active = this.telemetry.getActiveRuns().filter(r => r.investigationId === investigationId);
        return { investigation, stats, active };
    }

    close() {
        this.state.close();
    }
}

File: cyclops/core/engine.py
import asyncio
import json
from pathlib import Path

from cyclops.core.state import State
from cyclops.core.telemetry import Telemetry
from cyclops.connectors.registry import ConnectorRegistry
from cyclops.correlate.linker import Correlator
from cyclops.reporting.generator import Reporter

CONFIG_DIR = Path(__file__).parent.parent.parent / "config"


class Engine:
    def __init__(self, db_path=":memory:"):
        self.state = State(db_path)
        self.telemetry = Telemetry()
        self.registry = ConnectorRegistry(self.state, self.telemetry)
        self.correlator = Correlator(self.state, self.telemetry)
        self.reporter = Reporter(self.state)
        self.workflows = json.loads((CONFIG_DIR / "workflows.json").read_text())["workflows"]

    async def investigate(self, name, knowns, workflow=None):
        if not workflow:
            workflow = self._pick_workflow(knowns)
        wf = self.workflows.get(workflow)
        if not wf:
            raise ValueError(f"Unknown workflow: {workflow}")

        inv_id = self.state.create_investigation(name, workflow)
        for k in knowns:
            self.state.add_known(inv_id, k["type"], k["value"])

        self.state.update_investigation_status(inv_id, "running")

        phases = wf["phases"]
        completed_phases = set()

        while True:
            runnable = [
                p for p in phases
                if p["id"] not in completed_phases
                and all(d in completed_phases for d in p.get("depends_on", []))
            ]
            if not runnable:
                break

            for phase in runnable:
                await self._run_phase(inv_id, phase)
                completed_phases.add(phase["id"])

        self.state.update_investigation_status(inv_id, "completed")
        return inv_id

    async def _run_phase(self, inv_id, phase):
        phase_id = phase["id"]
        self.telemetry.phase_start(inv_id, phase_id)

        if phase.get("internal"):
            if phase_id == "correlation":
                self.correlator.correlate(inv_id)
            elif phase_id == "reporting":
                pass
            self.telemetry.phase_end(inv_id, phase_id, "completed")
            return

        connectors = self.registry.for_phase(phase)
        if not connectors:
            self.telemetry.phase_end(inv_id, phase_id, "completed")
            return

        input_types = phase.get("input_types", [])
        inputs = []
        for t in input_types:
            for k in self.state.get_knowns(inv_id, t):
                inputs.append((t, k["value"]))

        if not inputs:
            self.telemetry.phase_end(inv_id, phase_id, "completed")
            return

        tasks = []
        for entry in connectors:
            conn = entry["connector"]
            for input_type, input_value in inputs:
                if input_type in conn.config.get("accepts", []):
                    tasks.append(self._run_connector(conn, inv_id, phase_id, input_type, input_value))

        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)

        self.telemetry.phase_end(inv_id, phase_id, "completed")

    async def _run_connector(self, connector, inv_id, phase_id, input_type, input_value):
        try:
            return await connector.run(inv_id, phase_id, input_type, input_value)
        except Exception as e:
            return {"status": "failed", "error": str(e)}

    def _pick_workflow(self, knowns):
        types = {k["type"] for k in knowns}
        if types & {"domain", "ip"}:
            return "domain_recon"
        if types & {"email", "phone"}:
            return "person_full"
        return "username_trace"

    def close(self):
        self.state.close()
